logprep/util/auto_rule_corpus_tester.py

```diff
@@ -0,0 +1,483 @@
+
+    def run(self):
+        """
+        Starts the test routine by reading all input files, patching the logprep pipline, executing
+        the pipeline for each input event, comparing the generated output with the expected output
+        and printing out the test results.
+        """
+        self._read_files()
+        self._create_patched_pipeline()
+        self._run_logprep_per_test_case()
+        self._compare_with_expected_outputs()
+        self._print_detailed_reports()
+        self._print_test_statistics()
+        shutil.rmtree(self.tmp_dir)
+        if self.at_least_one_failed:
+            sys.exit(1)
+        else:
+            sys.exit(0)
+
+    def _read_files(self):
+        """Traverse the given input directory and find all test cases."""
+        data_directory = self.input_test_data_path
+        file_paths = self._collect_test_case_file_paths(data_directory)
+        test_cases = self._group_path_by_test_case(data_directory, file_paths)
+        self.test_cases = dict(sorted(test_cases.items()))
+
+    def _group_path_by_test_case(self, data_directory, file_paths):
+        test_cases = {}
+        for filename in file_paths:
+            test_case_id = self._strip_input_file_type(filename)
+            if test_case_id not in test_cases:
+                test_cases[test_case_id] = {
+                    "test_data_path": {"in": "", "out": "", "out_extra": ""},
+                    "report_print_statements": [],
+                }
+            data_path = test_cases.get(test_case_id, {}).get("test_data_path", {})
+            if "_in.json" in filename:
+                data_path.update({"in": os.path.join(data_directory, filename)})
+            if "_out.json" in filename:
+                data_path.update({"out": os.path.join(data_directory, filename)})
+            if "_out_extra.json" in filename:
+                data_path.update({"out_extra": os.path.join(data_directory, filename)})
+            test_cases[test_case_id]["test_data_path"].update(data_path)
+        return test_cases
+
+    def _collect_test_case_file_paths(self, data_directory):
+        file_paths = []
+        for root, _, files in os.walk(data_directory):
+            for filename in files:
+                file_paths.append(os.path.abspath(os.path.join(root, filename)))
+        return file_paths
+
+    def _run_logprep_per_test_case(self):
+        """
+        For each test case the logprep connector files are rewritten (only the current test case
+        will be added to the input file), the pipline is run and the outputs are compared.
+        """
+        for current_test_case in self.test_cases.items():
+            if not current_test_case[1].get("test_data_path", {}).get("in"):
+                raise ValueError(
+                    f"The test case '{current_test_case[0]}' is missing an input file."
+                )
+            self._prepare_connector_files(current_test_case)
+            self.pipeline._process_pipeline()
```

Comment:
I'm not fine with this. It couples the implementation details of the pipeline to this class. The method promised to run logprep per testcase, but it doesn't do it. Why not start logprep in a subprocess here like in the acceptance tests?

Reply:
Due to performance reasons I left this direct pipeline call. Sadly this means that I have to handle and reformat the logprep extra outputs to align them to the desired format. Calling logprep per test case in a subprocess would take to much time as every test case would mean that logprep would have to start completely again.